### PR TITLE
Initial version of ldap authentication plugin.

### PR DESCRIPTION
**Pull request by petervarkoly**
I've enhanced Radicale with ldap authentication plugin.
I've tested it and works fine with openldap2 and samba-ad.
In a next step I would like to enhance the rights module:
The ldap plugin reads the group of the authenticated user. From this reason I would enhance the from_file.py to be able handle group based rights.

### CODE REVIEW

**Comment by pbiering** [comment]
@petervarkoly : can you please create an empty commit to trigger "actions"? See here for more information: https://stackoverflow.com/questions/52408592/how-to-relaunch-github-check-without-pushing-new-commits

Also reduce the patch to minimal required changes and please do not apply changes to `config` and `rights` in an active way.

File: config [hunk 1]
```diff
@@ -15,7 +15,7 @@
 # IPv4 syntax: address:port
 # IPv6 syntax: [address]:port
 # For example: 0.0.0.0:9999, [::]:9999
-#hosts = localhost:5232
+hosts = 0.0.0.0:5232
```
Comment (pbiering) [inline config:18]: please remove this change from PR

File: config [hunk 2]
```diff
@@ -77,7 +92,7 @@
 #type = owner_only
 
 # File for rights management from_file
-#file = /etc/radicale/rights
+file = /etc/radicale/rights
```
Comment (pbiering) [inline config:95]: please remove this change from PR

File: radicale/app/__init__.py [hunk 3]
```diff
@@ -237,6 +237,11 @@ def response(status: int, headers: types.WSGIResponseHeaders,
                     authorization.encode("ascii"))).split(":", 1)
 
         user = self._auth.login(login, password) or "" if login else ""
+        try:
+            logger.debug("Groups %r",",".join(self._auth._ldap_groups))
+            self._rights._user_groups = self._auth._ldap_groups
+        except AttributeError:
+            pass
```
Comment (pbiering) [inline radicale/app/__init__.py:244]: shouldn't this be only active in case of "ldap" is used for authentication?

File: rights [hunk 4]
```diff
@@ -1,5 +1,34 @@
...
+groups: sysadmins,teachers,administration,administrators
+collection: {user}/[^/]+
+permissions: rw
+
+# Allow reading calendars and address books that are direct
+# children of the principal collection for other users
+[calendarsReader]
+user: .+
+collection: {user}/[^/]+
+permissions: r
```
Comment (pbiering) [inline rights:31]: please keep this all commented out as it is today and reduce to additional examples which can be used by "ldap"

File: config [hunk 5]
```diff
@@ -53,6 +53,24 @@
 [auth]
 
 # Authentication method
+# Value: none | htpasswd | remote_user | http_x_remote_user | ldap
```
Comment (pbiering) [inline config:56]: this is redundant with line 56, please move the following lines below `#type = none`

File: config [hunk 6]
```diff
@@ -177,4 +195,4 @@
 
 # When returning a free-busy report, limit the number of returned
 # occurences per event to prevent DOS attacks.
-#max_freebusy_occurrence = 10000
```
Comment (pbiering) [inline config:180]: please restore original, potentially line feed lost or is missing

File: rights [hunk 7]
```diff
@@ -1,5 +1,34 @@
-# -*- mode: conf -*-
-# vim:ft=cfg
+# Allow all rights for the Administrator
```
Comment (pbiering) [inline rights:1]: please comment out all as "empty" should be default. Also shift below existing examples, but add a header like "for use with LDAP authentication"

**Comment by pbiering** [comment]
@petervarkoly : thank you, please check my additional comments.

BTW: is _ldaps_ supported out-of-the-box and which LDAP TLS truststore is in use? Potentially for TLS additional options for the TLS truststore can be helpful to be able to use a custom instead of a system one.

File: config (hunk 6 above)
Comment (petervarkoly) [reply config:180]: also done

**Comment by petervarkoly** [comment]
Hello Peter,

basicaly you have right. To be honest until now I've only used Radicale on our product CRANIX with the local samba-ad-server as LDAP-server. So TLS settings was not important for me.
I'll implemnet additional tls settings but only for ldap3 part if you agree.

Viele Grüße

Dipl. Ing. Péter Varkoly
EDV/IT Beratung
+49-15257491205
i.A. Cranix-Solutions GmbH
Steuernummer: 238/283/00067
Ust-ID: DE257896578
________________________________
Von: Peter Bieringer ***@***.***>
Gesendet: Montag, 26. August 2024 21:11
An: Kozea/Radicale ***@***.***>
Cc: Peter Varkoly ***@***.***>; Mention ***@***.***>
Betreff: Re: [Kozea/Radicale] Initial version of ldap authentication plugin. (PR #1218)


@petervarkoly<https://github.com/petervarkoly> : thank you, please check my additional comments.

BTW: is ldaps supported out-of-the-box and which LDAP TLS truststore is in use? Potentially for TLS additional options for the TLS truststore can be helpful to be able to use a custom instead of a system one.

**Comment by pbiering** [comment]
@petervarkoly : please rebase to latest upstream, then I will merge

**Comment by petervarkoly** [comment]
*
@petervarkoly : please rebase to latest upstream, then I will merge

Done

Dipl. Ing. Péter Varkoly
EDV/IT Beratung
+49-15257491205
i.A. Cranix-Solutions GmbH
Steuernummer: 238/283/00067
Ust-ID: DE257896578
________________________________
Von: Peter Bieringer ***@***.***>
Gesendet: Sonntag, 1. September 2024 17:27
An: Kozea/Radicale ***@***.***>
Cc: Peter Varkoly ***@***.***>; Mention ***@***.***>
Betreff: Re: [Kozea/Radicale] Initial version of ldap authentication plugin. (PR #1218)


@petervarkoly<https://github.com/petervarkoly> : please rebase to latest upstream, then I will merge

**Review by pbiering — APPROVED** [review]
ok

**Comment by pbiering** [comment]
@petervarkoly : `Test / lint` is not happy about your code, please improve

**Review by pbiering — APPROVED** [review]
ok

**Comment by pbiering** [comment]
@petervarkoly : _lint_ found next issue, please try to run local tests before next commit, see https://github.com/Kozea/Radicale/wiki/Development-Testing-Release

**Review by pbiering — APPROVED** [review]
thank you

**Comment by pbiering** [comment]
@petervarkoly : _flake8_ is now happy, but _mypy_ reports now issues

**Review by pbiering — APPROVED** [review]
thank you

**Comment by pbiering** [comment]
@petervarkoly : hmm, latest commit broke something

```
radicale/auth/__init__.py:55: in BaseAuth
    _ldap_groups: set[str] = set([])
E   TypeError: 'type' object is not subscriptable
```

please try to run the testsuite completly on your side before next commit

**Comment by petervarkoly** [comment]
This works only since python3.11

I've fiexed it for older versions.

Dipl. Ing. Péter Varkoly
EDV/IT Beratung
+49-15257491205
i.A. Cranix-Solutions GmbH
Steuernummer: 238/283/00067
Ust-ID: DE257896578
________________________________
Von: Peter Bieringer ***@***.***>
Gesendet: Mittwoch, 11. September 2024 12:56
An: Kozea/Radicale ***@***.***>
Cc: Peter Varkoly ***@***.***>; Mention ***@***.***>
Betreff: Re: [Kozea/Radicale] Initial version of ldap authentication plugin. (PR #1218)


@petervarkoly<https://github.com/petervarkoly> : hmm, latest commit broke something

radicale/auth/__init__.py:55: in BaseAuth
    _ldap_groups: set[str] = set([])
E   TypeError: 'type' object is not subscriptable


please try to run the testsuite completly on your side before next commit

**Review by pbiering — APPROVED** [review]
ok

**Comment by pbiering** [comment]
@petervarkoly : no luck so far, _lint_ plus _test_ reporting now again issues

**Review by pbiering — APPROVED** [review]
ok

**Comment by pbiering** [comment]
@petervarkoly : now _lint_ needs to make happy again.

**Review by pbiering — APPROVED** [review]
ok

**Comment by pbiering** [comment]
@petervarkoly : well done now, but coincidently via https://github.com/Kozea/Radicale/discussions/1564 I found that there is already another LDAP plugin available https://github.com/malmeloo/radicale3-ldap-auth which was forked from  older ones - were you aware about?

Would it make sense for any alignment before merge?

**Comment by pbiering** [comment]
@petervarkoly : I will merge your code and will also notify the other project that it would be better to extend the merged code instead of having a plugin in parallel.

**Comment by malmeloo** [comment]
Probably! I'll be honest, I already switched away from Radicale due to performance issues on my NAS, but it's great to see native LDAP support. 'My' plugin is just a slightly modified version in a fork chain of 4 or 5 other repositories ;)

I'll add a note to the readme to indicate native LDAP support, to prevent confusion.

**Pull request merged**